##  Mid Bootcamp Project

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
data = pd.read_excel("/Users/belma/Documents/IRONHACK/Lessons/Week_5/Mid_bootcamp_project/Data/regression_data.xls", engine="xlrd")

### Standardize the Data

#### To Do's:
- checking for column name standard
- checking on NAN Values
- checking duplicates in id column
- droping id column

In [3]:
data.head()

id       date  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520 2014-10-13         3       1.00         1180      5650     1.0   
1  6414100192 2014-12-09         3       2.25         2570      7242     2.0   
2  5631500400 2015-02-25         2       1.00          770     10000     1.0   
3  2487200875 2014-12-09         4       3.00         1960      5000     1.0   
4  1954400510 2015-02-18         3       2.00         1680      8080     1.0   

   waterfront  view  condition  ...  sqft_above  sqft_basement  yr_built  \
0           0     0          3  ...        1180              0      1955   
1           0     0          3  ...        2170            400      1951   
2           0     0          3  ...         770              0      1933   
3           0     0          5  ...        1050            910      1965   
4           0     0          3  ...        1680              0      1987   

   yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15   price  
0             0    98178  47.5112 -122.257           1340        5650  221900  
1          1991    98125  47.7210 -122.319           1690        7639  538000  
2             0    98028  47.7379 -122.233           2720        8062  180000  
3             0    98136  47.5208 -122.393           1360        5000  604000  
4             0    98074  47.6168 -122.045           1800        7503  510000  

[5 rows x 21 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   id             21597 non-null  int64         
 1   date           21597 non-null  datetime64[ns]
 2   bedrooms       21597 non-null  int64         
 3   bathrooms      21597 non-null  float64       
 4   sqft_living    21597 non-null  int64         
 5   sqft_lot       21597 non-null  int64         
 6   floors         21597 non-null  float64       
 7   waterfront     21597 non-null  int64         
 8   view           21597 non-null  int64         
 9   condition      21597 non-null  int64         
 10  grade          21597 non-null  int64         
 11  sqft_above     21597 non-null  int64         
 12  sqft_basement  21597 non-null  int64         
 13  yr_built       21597 non-null  int64         
 14  yr_renovated   21597 non-null  int64         
 15  zipcode        2159

##### As a result I cannot see any categorical colums in the dataset, the categorical columns are already encoded. 

In [5]:
data.shape

(21597, 21)

#### Checking  for Nan Values in every column

In [6]:
data.isnull().values.any()

False

In [7]:
data.isna().sum()

id               0
date             0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
price            0
dtype: int64

##### As a result, i see theat no NAN values are in the columns

#### Checking for unique values and duplicates in id column

In [8]:
data['id'].value_counts()
# date time has a good format, nothing to adjust

795000620     3
8910500150    2
7409700215    2
1995200200    2
9211500620    2
             ..
3649100387    1
2767603649    1
1446403617    1
5602000275    1
1523300157    1
Name: id, Length: 21420, dtype: int64

#### As a result i need to drop duplicates in id and then drop id column afterwards

In [9]:
data = data.drop_duplicates('id', keep='last')
data.shape

(21420, 21)

In [10]:
data = data.drop(['id'], axis=1)

### Exploring the data

#### Data exploration & Data pre-processing

#### To Do's
- checking the quantity of every column value
- Change data type in a column if needed 
- plot the correlations
- check the discrete and continious numerical colums7
- plot the categorical colums
- check for outliners
- replace / Change values, if needed for normalization

In [11]:
data.describe()

bedrooms     bathrooms   sqft_living      sqft_lot        floors  \
count  21420.000000  21420.000000  21420.000000  2.142000e+04  21420.000000   
mean       3.373950      2.118429   2083.132633  1.512804e+04      1.495985   
std        0.925405      0.768720    918.808412  4.153080e+04      0.540081   
min        1.000000      0.500000    370.000000  5.200000e+02      1.000000   
25%        3.000000      1.750000   1430.000000  5.040000e+03      1.000000   
50%        3.000000      2.250000   1920.000000  7.614000e+03      1.500000   
75%        4.000000      2.500000   2550.000000  1.069050e+04      2.000000   
max       33.000000      8.000000  13540.000000  1.651359e+06      3.500000   

         waterfront          view     condition         grade    sqft_above  \
count  21420.000000  21420.000000  21420.000000  21420.000000  21420.000000   
mean       0.007610      0.235107      3.410784      7.662792   1791.170215   
std        0.086903      0.767165      0.650035      1.171971    828.692965   
min        0.000000      0.000000      1.000000      3.000000    370.000000   
25%        0.000000      0.000000      3.000000      7.000000   1200.000000   
50%        0.000000      0.000000      3.000000      7.000000   1560.000000   
75%        0.000000      0.000000      4.000000      8.000000   2220.000000   
max        1.000000      4.000000      5.000000     13.000000   9410.000000   

       sqft_basement      yr_built  yr_renovated      zipcode           lat  \
count   21420.000000  21420.000000  21420.000000  21420.00000  21420.000000   
mean      291.962418   1971.092997     84.793091  98077.87437     47.560197   
std       442.875502     29.387141    402.574626     53.47748      0.138589   
min         0.000000   1900.000000      0.000000  98001.00000     47.155900   
25%         0.000000   1952.000000      0.000000  98033.00000     47.471200   
50%         0.000000   1975.000000      0.000000  98065.00000     47.572100   
75%       560.000000   1997.000000      0.000000  98117.00000     47.678100   
max      4820.000000   2015.000000   2015.000000  98199.00000     47.777600   

               long  sqft_living15     sqft_lot15         price  
count  21420.000000   21420.000000   21420.000000  2.142000e+04  
mean    -122.213784    1988.384080   12775.718161  5.418614e+05  
std        0.140791     685.537057   27345.621867  3.675569e+05  
min     -122.519000     399.000000     651.000000  7.800000e+04  
25%     -122.328000    1490.000000    5100.000000  3.249500e+05  
50%     -122.230000    1840.000000    7620.000000  4.505500e+05  
75%     -122.125000    2370.000000   10086.250000  6.450000e+05  
max     -121.315000    6210.000000  871200.000000  7.700000e+06

In [12]:
data.head()

date  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0 2014-10-13         3       1.00         1180      5650     1.0           0   
1 2014-12-09         3       2.25         2570      7242     2.0           0   
2 2015-02-25         2       1.00          770     10000     1.0           0   
3 2014-12-09         4       3.00         1960      5000     1.0           0   
4 2015-02-18         3       2.00         1680      8080     1.0           0   

   view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
0     0          3      7        1180              0      1955             0   
1     0          3      7        2170            400      1951          1991   
2     0          3      6         770              0      1933             0   
3     0          5      7        1050            910      1965             0   
4     0          3      8        1680              0      1987             0   

   zipcode      lat     long  sqft_living15  sqft_lot15   price  
0    98178  47.5112 -122.257           1340        5650  221900  
1    98125  47.7210 -122.319           1690        7639  538000  
2    98028  47.7379 -122.233           2720        8062  180000  
3    98136  47.5208 -122.393           1360        5000  604000  
4    98074  47.6168 -122.045           1800        7503  510000

In [13]:
data.max()

date             2015-05-27 00:00:00
bedrooms                          33
bathrooms                        8.0
sqft_living                    13540
sqft_lot                     1651359
floors                           3.5
waterfront                         1
view                               4
condition                          5
grade                             13
sqft_above                      9410
sqft_basement                   4820
yr_built                        2015
yr_renovated                    2015
zipcode                        98199
lat                          47.7776
long                        -121.315
sqft_living15                   6210
sqft_lot15                    871200
price                        7700000
dtype: object

In [14]:
data.min()

date             2014-05-02 00:00:00
bedrooms                           1
bathrooms                        0.5
sqft_living                      370
sqft_lot                         520
floors                           1.0
waterfront                         0
view                               0
condition                          1
grade                              3
sqft_above                       370
sqft_basement                      0
yr_built                        1900
yr_renovated                       0
zipcode                        98001
lat                          47.1559
long                        -122.519
sqft_living15                    399
sqft_lot15                       651
price                          78000
dtype: object

##### As an outcoume, a beedroom size of 33 looks different than the rest, I will check on the single row. 

In [15]:
data[data['bedrooms'] == 33]

date  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
15856 2014-06-25        33       1.75         1620      6000     1.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
15856           0     0          5      7        1040            580   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
15856      1947             0    98103  47.6878 -122.331           1330   

       sqft_lot15   price  
15856        4700  640000

#### The sqmft size of 1620  is unrealistc for 33 bedrooms with the combination of the 1,75 bathrooms in relation. , so I will replace the value with 3, since this seems to be a typo. 

In [16]:
data['bedrooms'] = data['bedrooms'].replace(33, 3)  
data[data['bedrooms'] == 33]

Empty DataFrame
Columns: [date, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15, price]
Index: []

In [17]:
data.to_csv('cleaned_data.csv', index=False)